db/view/view.cc

```diff
@@ -2413,13 +2415,55 @@ void view_updating_consumer::maybe_flush_buffer_mid_partition() {
     }
 }
 
+// lives in a seastar thread
+class view_update_pusher {
+    lw_shared_ptr<replica::table> _table;
+    std::vector<sstables::shared_sstable> _excluded_sstables;
+    std::unique_ptr<const dht::partition_range> _curr_pr;
+    flat_mutation_reader_v2_opt _reader;
+
+private:
+    mutation_source as_mutation_source() {
+        return mutation_source([this] (
+                schema_ptr s,
+                reader_permit permit,
+                const dht::partition_range& range,
+                const query::partition_slice& slice,
+                const io_priority_class& pc,
+                tracing::trace_state_ptr trace_state,
+                streamed_mutation::forwarding,
+                mutation_reader::forwarding) -> flat_mutation_reader_v2 {
+            if (!_reader) {
+                _curr_pr = std::make_unique<dht::partition_range>(range);
+                _reader = _table->make_reader_v2_excluding_sstables(s, permit, _excluded_sstables, *_curr_pr, s->full_slice(), pc,
+                        std::move(trace_state), streamed_mutation::forwarding::yes, mutation_reader::forwarding::yes);
+            // Check if we are continuing from the last partition or not.
+            } else if (!range.start()->value().equal(*s, _curr_pr->start()->value())) { // pr is always singular
+                auto prev_pr = std::exchange(_curr_pr, std::make_unique<dht::partition_range>(range));
+                _reader->fast_forward_to(*_curr_pr).get();
+            }
+            return make_slicing_forwarding_reader(make_delegating_reader(*_reader), slice);
```

Comment:
Did you mean to ask here? Indeed here, if we read twice from the same reader, the second instance will not have a partition header.

Reply:
No, I meant there, but was confused.